Mathlib/Analysis/SpecialFunctions/Log/ENNRealLogExp.lean

```diff
@@ -0,0 +1,170 @@
+lemma exp_lt_exp_iff {a b : EReal} : exp a < exp b ↔ a < b := by
+  conv_rhs => rw [← log_exp a, ← log_exp b, log_lt_log_iff]
+
+@[simp] lemma zero_lt_exp_iff {a : EReal} : 0 < exp a ↔ ⊥ < a := exp_bot ▸ @exp_lt_exp_iff ⊥ a
+
+@[simp] lemma exp_lt_top_iff {a : EReal} : exp a < ⊤ ↔ a < ⊤ := exp_top ▸ @exp_lt_exp_iff a ⊤
+
+@[simp] lemma exp_lt_one_iff {a : EReal} : exp a < 1 ↔ a < 0 := exp_zero ▸ @exp_lt_exp_iff a 0
+
+@[simp] lemma one_lt_exp_iff {a : EReal} : 1 < exp a ↔ 0 < a := exp_zero ▸ @exp_lt_exp_iff 0 a
+
+@[simp] lemma exp_le_exp_iff {a b : EReal} : exp a ≤ exp b ↔ a ≤ b := by
+  conv_rhs => rw [← log_exp a, ← log_exp b, log_le_log_iff]
+
+@[simp] lemma exp_le_one_iff {a : EReal} : exp a ≤ 1 ↔ a ≤ 0 := exp_zero ▸ @exp_le_exp_iff a 0
+
+@[simp] lemma one_le_exp_iff {a : EReal} : 1 ≤ exp a ↔ 0 ≤ a := exp_zero ▸ @exp_le_exp_iff 0 a
+
+lemma exp_nmul (x : EReal) (n : ℕ) : exp (n * x) = (exp x) ^ n := by
+  simp_rw [← log_eq_iff, log_pow, log_exp]
+  norm_cast
+
+lemma exp_mul (x : EReal) (y : ℝ) : exp (x * y) = (exp x) ^ y := by
+  rw [← log_eq_iff, log_rpow, log_exp, log_exp, mul_comm]
+
+end Exp
+
+section OrderIso
+
+/-- `ENNReal.log` and its inverse `ENNReal.exp` are an order isomorphism between `ℝ≥0∞` and
+`EReal`. -/
+noncomputable
+def logOrderIso : ℝ≥0∞ ≃o EReal where
+  toFun := log
+  invFun := exp
+  left_inv x := exp_log x
+  right_inv x := log_exp x
+  map_rel_iff' := by simp only [Equiv.coe_fn_mk, log_le_log_iff, forall_const]
+
+@[simp] lemma logOrderIso_apply (x : ℝ≥0∞) : logOrderIso x = log x := rfl
+@[simp] lemma logOrderIso_symm_apply (x : EReal) : logOrderIso.symm x = exp x := rfl
+
+end OrderIso
+
+section Continuity
+
+/-- `log` as a homeomorphism. -/
+noncomputable def log_homeomorph : ℝ≥0∞ ≃ₜ EReal := logOrderIso.toHomeomorph
+
+@[simp] lemma log_homeomorph_apply (x : ℝ≥0∞) : log_homeomorph x = log x := rfl
+
+/-- `exp` as a homeomorphism. -/
+noncomputable def exp_homeomorph : EReal ≃ₜ ℝ≥0∞ := logOrderIso.symm.toHomeomorph
+
+@[simp] lemma exp_homeomorph_apply (x : EReal) : exp_homeomorph x = exp x := rfl
+
+@[continuity, fun_prop]
+lemma continuous_log : Continuous log := logOrderIso.continuous
+
+@[continuity, fun_prop]
+lemma continuous_exp : Continuous exp := logOrderIso.symm.continuous
+
+end Continuity
+
+section Measurability
+
+@[measurability, fun_prop]
+lemma measurable_log : Measurable log := continuous_log.measurable
+
+@[measurability, fun_prop]
+lemma measurable_exp : Measurable exp := continuous_exp.measurable
+
+@[measurability, fun_prop]
+lemma _root_.Measurable.ereal_log {α : Type*} {_ : MeasurableSpace α}
+    {f : α → ℝ≥0∞} (hf : Measurable f) :
+    Measurable fun x ↦ log (f x) := measurable_log.comp hf
+
+@[measurability, fun_prop]
+lemma _root_.Measurable.ereal_exp {α : Type*} {_ : MeasurableSpace α}
+    {f : α → EReal} (hf : Measurable f) :
+    Measurable fun x ↦ exp (f x) := measurable_exp.comp hf
+
+end Measurability
+
+end ENNReal
+
+instance : TopologicalSpace.MetrizableSpace EReal :=
```

Comment:
Could you even register a polish space instance?

Reply:
We are not sure how to do it. @LorenzoLuccioli

Reply:
I just committed an attempt at adding the `PolishSpace` instance: 
https://github.com/leanprover-community/mathlib4/blob/84f997b1aafbf27bda3206e2ff634b5b633b1ff3/Mathlib/Analysis/SpecialFunctions/Log/ENNRealLogExp.lean#L242-L249

I was missing the instance `(uniformity EReal).IsCountablyGenerated` and I tried to prove it adding the `PseudoEMetricSpace` instance to `EReal`. The problem now is that Lean does not recognize that the `UniformSpace` structure induced by the `EDist` and the one induced by the topology are the same, do you know how to prove this?

Moreover I'm not sure whether my implementation of `EDist` satisfies the best practices, and I think it should be moved to some other file, I just left it here temporarily.

Reply:
I solved the problem with the help of @RemyDegenne, proving that `ENNReal` is a `PolishSpace` since it is isomorphic to a real interval, then we recover the property for `EReal` through the new isomorphism with `ENNReal`. Since it was not needed anymore I erased the definition of `EDist`.

Now I just have to figure out where to position the `PolishSpace` instance for `ENNReal`. 
We cannot put it in the ENNReal [file](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Instances/EReal.html), since it is imported by the file where `PolishSpace` is defined.
We could leave it here, but it seems weird to do that, since it does not really use anything about the log or exp.
`find_home` suggests to put it in the Polish spaces [file](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/MetricSpace/Polish.html), this would avoid any additional import, but in that file there are no instances of specific types to be Polish spaces.
Do you have any suggestions?

Reply:
I think it would make sense to put it in the file defining PolishSpace, since it can not be put earlier.

Note that if you register the PolishSpace instance, you can remove the metrizable one as it follows from it.